### PR TITLE
Security Enhancement: Distance Check for 'esx:onPickup' Event

### PR DESCRIPTION
**Pull request by Kenshiin13**
#### Description

This pull request introduces a critical security update for the "esx:onPickup" event to mitigate potential abuses. The update incorporates a distance check mechanism, preventing players from exploiting the event to collect pickups from excessive distances. This measure ensures fair gameplay and maintains the integrity of the event.

#### Changes Made

- Store serverside pickup coords as a vector, eliminating the need for client-side conversion, and allowing for arithmetic operations on them. 
- Implemented a distance check within the existing 'esx:onPickup' event handling code.
- If a player attempts to interact with a pickup beyond a specified distance (5.0 units), a warning is triggered, and the interaction is blocked.

The modified code snippet is as follows:
```lua
RegisterNetEvent('esx:onPickup')
AddEventHandler('esx:onPickup', function(pickupId)
    local pickup, xPlayer, success = Core.Pickups[pickupId], ESX.GetPlayerFromId(source)

    if pickup then
        local playerPickupDistance = #(pickup.coords - xPlayer.getCoords(true))
        if(playerPickupDistance > 5.0) then
            print(('[^3WARNING^7] Player Detected Cheating (Out of range pickup): ^5%s^7'):format(xPlayer.getIdentifier()))
            return
        end

        -- Rest of the original event handling code

        if success then
            Core.Pickups[pickupId] = nil
            TriggerClientEvent('esx:removePickup', -1, pickupId)
        end
    end
end)
```

#### Reasoning

This security update is vital to prevent potential abuses of the "esx:onPickup" event. The event could potentially be exploited by malicious players to trigger unauthorized pickups using a simple code snippet like this:
```lua
RegisterNetEvent("esx:createPickup", function(pickupId)
    TriggerServerEvent("esx:onPickup", pickupId)
end)
 ```
By implementing a distance check, the update ensures that players can only interact with pickups within a reasonable range, discouraging unauthorized collection of pickups from afar.

#### Counterargument Consideration:

It's worth addressing the potential counterargument that the distance check might be unnecessary due to a player's ability to teleport to pickups, thus eliminating the need for this change. However, it's essential to recognize that third-party anticheats are designed precisely to identify players employing teleportation tactics across the map. Additionally, the presence of one form of vulnerability should not overshadow addressing another potential attack vector.

Thank you for considering these enhancements. Your feedback and suggestions are appreciated.

### CODE REVIEW

**Review by iSentrie — CHANGES_REQUESTED** [review]
this will require decrease in vector z, else player won't be able to pick anything from ground

```lua
TriggerEvent('esx:createPickup', pickupId, pickup.label, pickup.coords - vector3(0, 0, 0.5), pickup.type, pickup.name, pickup.components, pickup.tintIndex)
```

however, who don't use ox_inventory, after client restart missing pickups will appear floating in the air, needs an better fix, but for now at least it will work. currently in last version missing pickups wasn't working at all.

**Comment by Kenshiin13** [comment]
> this will require decrease in vector z, else player won't be able to pick anything from ground

Haven't tested that tbh. The pickups spawned properly on the ground when I tested the vulnerability. I can push another commit tho if requested.

**Comment by iSentrie** [comment]
i tested, that's why i wrote it

**Comment by Kenshiin13** [comment]
> your changes fixes the error

My changes have nothing to do with that.

**Comment by iSentrie** [comment]
> > your changes fixes the error
> 
> My changes have nothing to do with that.

without your changes missing pickups won't spawn at all, resulting in an error in client console.
you fixed it without knowing it.
it was fixed before by someone and then broken again by someone, now you fixed it again.

**Comment by Gellipapa** [comment]
@Kenshiin13 Hi! Could you write in to accept coordinates not just the current coordinate? 
Example: I need to tell you that I want a pickup points xyz, because currently it only looks at your current coordinate to the best of my knowledge.

**Comment by Kenshiin13** [comment]
> @Kenshiin13 Hi! Could you write in to accept coordinates not just the current coordinate? Example: I need to tell you that I want a pickup points xyz, because currently it only looks at your current coordinate to the best of my knowledge.

I'm sorry, I'm not sure what you mean. Why would we test against anything other than the players current coordinates to determine whether a pickup request ist legitmate or not?

**Comment by Gellipapa** [comment]
> 
I seem to remember that currently in ESX there is no option to spawn a pickup point to a given coordinate, you can only spawn a point to the player's coordinate. That's what I was thinking about, that I should be able to spawn a pickup point to a custom coordinate.

**Comment by Kenshiin13** [comment]
> > 
> 
> I seem to remember that currently in ESX there is no option to spawn a pickup point to a given coordinate, you can only spawn a point to the player's coordinate. That's what I was thinking about, that I should be able to spawn a pickup point to a custom coordinate.

You mean like this 92d6f49464fbffc2f5431c06f560637eb457302a?

**Comment by Gellipapa** [comment]
@Kenshiin13 Yes, thank you.

**Comment by Gellipapa** [comment]
Thanks for the PR, we will test it soon and if it works we will merge it.

**Comment by Gellipapa** [comment]
@Kenshiin13 Hi! I tested this pr and i found bug.

- ESX.CreatePickup method first parameter is type (this is not correct name, please fix) and this is name overwrite default lua type checker and i get nil value if i try run this code. (Please rename all type parameter to other name in server/main.lua)

Let's find a solution so that if I don't want to spawn a weapon, I don't have to specify "components" and "tintIdex", because I currently have to specify them because coordinates are included and the parameter number of pieces makes it mandatory to specify them.  **(!!We have to pay attention to backwards compatibility because it is not easy to do this.)**


image

File: [core]/es_extended/server/functions.lua [hunk 1]
```diff
@@ -469,10 +469,10 @@ function ESX.GetUsableItems()
 end
 
 if not Config.OxInventory then
-	function ESX.CreatePickup(type, name, count, label, playerId, components, tintIndex)
+	function ESX.CreatePickup(type, name, count, label, playerId, components, tintIndex, coords)
```
Comment (Gellipapa) [inline [core]/es_extended/server/functions.lua:472]: Rename type parameter because overwrite default lua type checker method.

**Review by Gellipapa — CHANGES_REQUESTED** [review]

**Comment by Gellipapa** [comment]
In the future, please make sure that it is tested, because this is a basic bug that could have been fixed in the first test. Thanks.

**Comment by Kenshiin13** [comment]
> Let's find a solution so that if I don't want to spawn a weapon, I don't have to specify "components" and "tintIdex", because I currently have to specify them because coordinates are included and the parameter number of pieces makes it mandatory to specify them. **(!!We have to pay attention to backwards compatibility because it is not easy to do this.)**
> 
> image

Well, I am not sure if there is a way to do this without breaking backwards-compatibility.

**Comment by Gellipapa** [comment]
> > Let's find a solution so that if I don't want to spawn a weapon, I don't have to specify "components" and "tintIdex", because I currently have to specify them because coordinates are included and the parameter number of pieces makes it mandatory to specify them. **(!!We have to pay attention to backwards compatibility because it is not easy to do this.)**
> 
> Well, I am not sure if there is a way to do this without breaking backwards-compatibility.

If you have any ideas, feel free to write them down, I can't think of anything right now, if we don't have a solution, it will stay as it was.

**Comment by Kenshiin13** [comment]
> If you have any ideas, feel free to write them down, I can't think of anything right now, if we don't have a solution, it will stay as it was. I mean that part will not be changed.

Nope, not really. Does that work for you? https://github.com/esx-framework/esx_core/pull/1161/commits/1b27475ce16abbee6aeec606f9bf123069b5f081

File: [core]/es_extended/server/functions.lua (hunk 1 above)
Comment (Kenshiin13) [reply [core]/es_extended/server/functions.lua]: https://github.com/esx-framework/esx_core/pull/1161/commits/1b27475ce16abbee6aeec606f9bf123069b5f081

**Comment by Gellipapa** [comment]
> > If you have any ideas, feel free to write them down, I can't think of anything right now, if we don't have a solution, it will stay as it was. I mean that part will not be changed.
> 
> Nope, not really. Does that work for you? [1b27475](https://github.com/esx-framework/esx_core/commit/1b27475ce16abbee6aeec606f9bf123069b5f081)

I will check soon.

**Comment by Gellipapa** [comment]
Hi! Well done, I was tested and merged.

**Review by Gellipapa — APPROVED** [review]

**Pull request merged**